File: backend/app/validators/emergency_validator.py
from ..models.emergency import EmergencyModel
from ..schemas.emergency_request import EmergencyRequestCreate

class EmergencyValidationError(Exception):
    pass

class EmergencyValidator:
    @staticmethod
    def validate_creation(request_data: EmergencyRequestCreate):
        if not request_data.user_id:
            raise EmergencyValidationError("User ID is required to create an emergency.")
        if request_data.location.latitude == 0 and request_data.location.longitude == 0:
            raise EmergencyValidationError("Valid location coordinates are required.")

    @staticmethod
    def validate_cancellation(emergency: EmergencyModel):
        if not emergency:
            raise EmergencyValidationError("Emergency request does not exist.")
        if emergency.status == "completed":
            raise EmergencyValidationError("Cannot cancel a completed emergency.")
        if emergency.status == "cancelled":
            raise EmergencyValidationError("Emergency is already cancelled.")

File: frontend/lib/features/emergency/services/emergency_validator.dart
import '../domain/entities/emergency_request.dart';
import '../domain/enums/emergency_status.dart';

class EmergencyValidator {
  static String? validateRequest(EmergencyRequest request) {
    if (request.userId.isEmpty) {
      return 'User ID is required';
    }
    // Location validation (already strongly typed as non-nullable, but can check logical coordinates if needed)
    if (request.location.latitude == 0 && request.location.longitude == 0) {
      // In a real app, 0,0 might be valid (Null Island), but typically indicates failure to fetch
      return 'Valid location coordinates are required';
    }
    return null;
  }

  static String? validateTransition(EmergencyStatus currentStatus, EmergencyStatus nextStatus) {
    if (currentStatus == EmergencyStatus.completed && nextStatus != EmergencyStatus.completed) {
      return 'Cannot transition from COMPLETED to any other status';
    }
    if (currentStatus == EmergencyStatus.cancelled) {
      return 'Cannot transition from CANCELLED';
    }
    if (nextStatus == EmergencyStatus.cancelled && currentStatus == EmergencyStatus.completed) {
      return 'Cannot cancel a COMPLETED request';
    }
    return null;
  }
}
